### ISSUE TITLE
Cannot use Abstract Refinements to Refine Monad Instances

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#1057]
I cannot figure out why the below is failing; it is very similar to `benchmarks/icfp15/pos/RIO2.hs` which  (presumably) works. The error is in the monad instance, specifically:

```
rjhala@borscht ~/r/s/liquid-sf (master)> stack exec -- liquid State0.hs 

 /Users/rjhala/research/sf/liquid-sf/State0.hs:50:10-17: Error: Uh oh.
 
 50 | instance Monad ST where
               ^^^^^^^^
     Predicate in lhs of constraint:lq_tmp$x##3096:(ST a) -> lq_tmp$x##3097:(lq_tmp$x##3094:a -> (ST b)) -> (ST b)
<:
lq_tmp$x##3047:(ST a) -> lq_tmp$x##3048:(lq_tmp$x##3046:a -> (ST b)) -> (ST b)
```

Say what?!

```haskell
{-@ LIQUID "--no-pattern-inline" @-}

module State (
   returnST -- :: a -> ST a s
 , bindST   -- :: ST a s -> (a -> ST b s) -> ST b s
 , ST(..)
 ) where

import Prelude hiding (snd, fst)

type State = Int

data ST a = S (State -> (a, State))
{-@ data ST a <pre :: State -> Bool, post :: a -> State -> Bool>
       = S (ys::(x:State<pre> -> ((a, State)<post>)))
  @-}

{-@ returnST :: forall <p :: a -> State -> Bool>.
                      x:a -> ST <{v:State<p x>| true}, p> a
  @-}
returnST :: a -> ST a
returnST x = S $ \s -> (x, s)

{-@ bindST :: forall <p :: State -> Bool, q :: a -> State -> Bool, r :: b -> State -> Bool>.
                         ST <p, q> a
                      -> (x:a -> ST <{v:State<q x> | true}, r> b)
                      -> ST <p, r> b
  @-}
bindST :: ST a -> (a -> ST b) -> ST b
bindST (S m) k = S $ \s -> let (a, s') = m s in apply (k a) s'

{-@ apply :: forall <p :: State -> Bool, q :: a -> State -> Bool>.
             ST <p, q> a -> State<p> -> (a, State)<q>
  @-}
apply :: ST a -> State -> (a, State)
apply (S f) s = f s

--------------------------------------------------------------------------------

instance Functor ST where
  fmap  = undefined

instance Applicative ST where
  pure  = undefined
  (<*>) = undefined

instance Monad ST where
  {-@ instance Monad ST where
      return :: forall <p :: a -> State -> Bool>.
                  x:a -> ST <{v:State<p x>| true}, p> a
      ;
      >>=    :: forall <p :: State -> Bool, q :: a -> State -> Bool, r :: b -> State -> Bool>.
                               ST <p, q> a
                            -> (x:a -> ST <{v:State<q x> | true}, r> b)
                            -> ST <p, r> b 
    @-}
  return = returnST
  (>>=)  = bindST

---

{-@ fresh :: ST <{\v -> (0 <= v)}, {\rv v -> ( 0 <= rv && rv + 1 = v )}> Int @-}
fresh :: ST Int
fresh = S (\n -> (n, n+1))

{-@ incr2 :: ST <{\v -> (0 <= v)}, {\rv v -> (0 <= rv && 1 <= v)}> Int @-}
incr2 :: ST Int
incr2 = fresh `bindST` returnST
```

**Comment by ranjitjhala** [comment]
See #1071

**Comment by ranjitjhala** [comment]
Fixed by @nikivazou in #1071
